libraries/AP_HAL_PX4/RCOutput.cpp

```diff
@@ -495,8 +469,7 @@ void PX4RCOutput::_send_outputs(void)
                             ap_uc->rco_write(_period[i], i);
                         }
 
-                        bool armed = hal.util->get_soft_armed();
-                        if (armed) {
+                        if (hal.util->safety_switch_state() != AP_HAL::Util::SAFETY_DISARMED) {
```

Comment:
> Well, hal.util->get_soft_armed(); is actually used a lot in code to determine armed state (in very many places of vehicle specific code).

I'll quote myself:

> The only reason soft armed exist in the HAL is because libraries needed to know if vehicles had armed or not.

So yes, libraries use soft armed to determine if the vehicle is armed. Some libraries also the safety switch state. It all depends on their objective.
What I said is that no other HAL uses the soft armed state to determine if motors are armed or not.

> As for servos, in VRBrain RCOutput it is used also for same.

VRBrain is a dead HAL, it needs to be merged into the PX4 (and this renamed to NuttX). But not even that one uses soft armed state to determine if motors armed or not, it uses it to decide if it should 0 to CAN ESCs or the actual value - for the arming it uses the safety switch, it is a copy of what was done in PX4 before. As I said above, I'd be fine with the 0 thing, although I don't see the need for it - unless CAN ESCs don't stop when they don't receive any value?

> However I checked for hal.util->safety_switch_state() != AP_HAL::Util::SAFETY_DISARMED and it is also used a lot and actually by update_soft_armed() the hal.util->get_soft_armed(); is reassembling the safety switch but also adds arming.is_armed() check.
>
> So I believe correct will be: arming.is_armed() && hal.util->safety_switch_state() != AP_HAL::Util::SAFETY_DISARMED if we want to refuse from using get_soft_armed();

The `update_soft_armed` doesn't exist in Copter, nor does Copter use the `is_armed` method of the AP_Arming library. The soft armed state is defined by each vehicle as it pleases and the HAL layer shouldn't really care about it.
I'm pretty sure the current code doesn't work is motor test in Copter because it never tells the HAL it is armed. It could, but the copter should be on the bench, not ready to fly, so why would it be in armed state? As I said, it is up to the vehicle to decide this.

On my opinion once the RCOutput object in the HAL receives values to write it should indeed output them. If the safety switch is in disarmed state then it depends: do the ESCs support arm/disarm? Usual ESCs don't, so we stop sending PWM out, but CAN ESCs do, so we can send the value but tell them to stay disarmed.
To be completely honest this should be handled in method `force_safety_pending_requests` instead of here, but I'm not changing it without someone with CAN ESCs carefully testing it.

Ups, that was a testament, sorry :slightly_smiling_face: